### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by imahnama**
## In this Exercise are :

-Different methods, each containing built manually, including #each, #each_with_index , #select, #all?, #any?, #none?, #count, #inject, #map

### CODE REVIEW

File: README.md
Comment (majovanilla) [inline README.md:3]: Nice description!! :+1:

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- You are not supposed to use the original ruby methods but to create your own. In this case, you can use a while loop that finishes when you reached the last element of the enumerable.

- Your method is not returning an enumerator when no block is given. (You should move line 5 outside the loop, because you don't want to return every loop step).

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- You are not supposed to use the original ruby methods but to create your own. In this case, you can use a while loop that finishes when you reached the last element of the enumerable.
- Your method is not returning an enumerator when no block is given.

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Your method is not returning an enumerator when no block is given.

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains truthy values and false if it contains at least one falsy value.

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_all?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_all?(/w/)).`

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_all?(5)).`

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: You closed your enumerable module with this end. You should delete it and move it to the end.

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains at least one truthy value.

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_any?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_any?(/w/)).`

- Your method doesn't consider the possibility when a string is given as an argument `(some_enumerable.my_any?('cow')).`

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains no truthy value.

- Your method doesn't consider the possibility when a class is given as an argument `(some_enumerable.my_none?(String)).`

- Your method doesn't consider the possibility when a Regex is given as an argument `(some_enumerable.my_none?(/w/)).`

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_none?(8)).`

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no argument or block is given. It should return the count of all elements inside the enumerable.

- Your method doesn't consider the possibility when a number is given as an argument `(some_enumerable.my_count(5)).` It should return the count of all elements that match the argument.

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_map` **(covered by automated tests)** identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should return an enumerator if no block or argument is given.

File: enumarable_methods.rb
Comment (majovanilla) [inline enumarable_methods.rb]: - [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should work with symbols as arguments (:+) and perform the right operation.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hey @imahnama !!

## You did a *great* job on this project! 💜  And you are almost there!! :muscle: 

## Please, make all REQUIRED changes on the commented code below :point_down:  and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good Progress :+1: 
You have made the requested changes as suggested by the previous reviewer and that's good. However, we need to fix a few failing edge cases before we can merge.
Check my comments for suggestions

### Martin Afani

File: enumarable_methods.rb
Comment (whiz25) [inline enumarable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] counts the number of elements yielding a true value if a block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_count(&block) == array.count(&block)
```

File: enumarable_methods.rb
Comment (whiz25) [inline enumarable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given
```ruby
array = Array.new(100){rand(0...9)}
array.my_map == Enumerator #true
```

File: enumarable_methods.rb
Comment (whiz25) [inline enumarable_methods.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
array = Array.new(100) { rand(0...9) }
array.my_inject(:+) == array.inject(:+) #true
```
- [x] combines all elements of enum by applying a binary operation, specified by a block
```ruby
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word } 
array = Array.new(100){rand(0...9)}
words = %w[dog door rod blade]

array.my_inject(&operation) == array.inject(&operation) #true
actual = range.my_inject(4) { |prod, n| prod * n }
expected = range.inject(4) { |prod, n| prod * n }
actual == expected #true
words.my_inject(&search) == words.inject(&search) #true
```

File: enumarable_methods.rb
Comment (stratospherique) [inline enumarable_methods.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_count counts the number of elements yielding a true value if a block is given
```ruby
array = [1,2,8,9,7,5,6,3,11]
block = proc { |num| num < 5 }
array.my_count(&block) == array.count(&block) # false
```

File: enumarable_methods.rb
Comment (stratospherique) [inline enumarable_methods.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map == array.map # false
  ```

File: enumarable_methods.rb
Comment (stratospherique) [inline enumarable_methods.rb]: * [x] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```
* [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    array = [1,7,8,89,9,6,5,12,4,2,5]
    p array.my_inject(:+) == array.inject(:+) # return this error
    # LocalJumpError:
       no block given (yield)
     # ./enumarable_methods.rb:116:in `block in my_inject'
     # ./enumarable_methods.rb:8:in `my_each'
     # ./enumarable_methods.rb:116:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Progress @imahnama Almost there :clap: :clap: However, some methods need further refactoring in order to get approval for the project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods.

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Review by stratospherique — APPROVED** [review]
### Impressive work implementing the suggested changes :clap: :clap: Congrats, the project is approved :heavy_check_mark: You may merge :+1: :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

**Pull request merged**
